### PR TITLE
Chunking for expect/expect_and_grad

### PR DESCRIPTION
**Pull request by PhilipVinc**
This is an alternative implementation to #830, taking into account some of the feedback that was raised in that PR.

The implementation is much simpler and builds upon a `nk.jax.vmap_batched` that is semantically equivalent to `vmap` but uses scan under the hood to evaluate one batch at the time.

The limitation of minibatch size being a multiple of the number of chains is also relaxed. 
Now minibatch size should be an integer divisor of total number of samples per mpi rank, and attempts to achieve that by increasing the chain length.
This requirement is not mandatory and could be relaxed, and i agree that if you have a prime number of chains it might blow up the chain length, however respecting it leads to much shorter compilation times and slightly better performance.
So I am open to alternatives.

I'd like some feedback and someone to benchmark this a bit (I do not have the bandwith right now).

TODO: Make all operator types work with batching, make `vmaop_batched` autodiffable (@inailuig). docs

Credit for a lot of the batching logic goes to @inailuig who definitely was not tied to a chair and obliged to do this under duress.
This PR also contains a commit from his PR #912

### CODE REVIEW

**Comment by PhilipVinc** [comment]
@attila-i-szabo let me know if you find some time to play a bit with this if you have some feedback

**Comment by PhilipVinc** [comment]
This PR is essentially ready, even though tests are missing, as it now supports batching for both the forward and backward pass.
There are two PRs inside of it: @inailuig will soon open a PR with all the batching utilities that he contributes to `jax.vjp` (that right now are included within this PR) such as `nkjax.vjp_batched` and `nkjax.vmap_batched`.
Once that other PR is merged we can also hook up the QGT PR.

On the standard ising 1d with rbm model using a batch size of 128 incurs in a 9% slowdown wrt no batching. I suspect this is mainly overhead of the looping constructs.
Some more extensive benchmarks should be done. 
I would be curious @wdphy16 if you had some time to try this PR and see if it works well for RNN and allow you to avoid memory problems (to try, just checkout this branch and set `vstate.minibatch_size` to some value).

However, since batching is opt-in (setting `vstate.minibatch_size=xxx`) and since it allows to use very large number of samples, and it has to do some additional bookeping and operations, i think it will be hard to reduce the slowdown much.

The only additional thing that is missing would be a way to use minibatches in sampling (for example with autoregressive sampling, since it does it in only one go) but hopefully there is less of an issue there...
but that is left for another PR.

**Comment by wdphy16** [comment]
Good to hear that! I'll do some benchmarks in the next few days.

**Comment by inailuig** [comment]
> On the standard ising 1d with rbm model using a batch size of 128 incurs in a 9% slowdown wrt no batching. I suspect this is mainly overhead of the looping constructs.

Not sure what exactly you tested, but could it be the vjp which re-computes the forward pass?

**Comment by PhilipVinc** [comment]
Yes, most likely
Il 22 set 2021, 19:52 +0200, Clemens Giuliani ***@***.***>, ha scritto:

File: netket/vqs/mc/mc_state/expect_batched.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        σp = σp.reshape((σ.shape[0], -1, σ_shape[-1]))
+        mels = mels.reshape(σp.shape[:-1])
+
+    def logpsi(w, σ):
+        return model_apply_fun({"params": w, **model_state}, σ)
+
+    def log_pdf(w, σ):
+        return machine_pow * model_apply_fun({"params": w, **model_state}, σ).real
+
+    local_value_vmap = partial(local_value_kernel, logpsi, batch_size=batch_size)
```
Comment (inailuig) [inline netket/vqs/mc/mc_state/expect_batched.py:104]:
```suggestion
    local_value_vmap = partial(local_value_kernel, logpsi, chunk_size=batch_size)
```

Comment (attila-i-szabo) [reply netket/vqs/mc/mc_state/expect_batched.py:104]: change all occurrences of `batch` to `chunk`? or are there contexts where we still want it to be `batched`?

File: netket/vqs/mc/mc_state/expect_grad_batched.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    n_samples = σ.shape[0] * mpi.n_nodes
+
+    O_loc = local_value_kernel_chunked(
+        model_apply_fun,
+        {"params": parameters, **model_state},
+        σ,
+        σp,
+        mels,
+        batch_size=batch_size,
```
Comment (attila-i-szabo) [inline netket/vqs/mc/mc_state/expect_grad_batched.py:116]: `jax._src.traceback_util.UnfilteredStackTrace: TypeError: local_valuekernel_chunked() got an unexpected keyword argument 'batch_size'`

Comment (PhilipVinc) [reply netket/vqs/mc/mc_state/expect_grad_batched.py]: addressed

File: netket/vqs/mc/mc_state/expect_batched.py (hunk 1 above)
Comment (PhilipVinc) [reply netket/vqs/mc/mc_state/expect_batched.py]: done

**Comment by PhilipVinc** [comment]
I think this is ready to be reviewed.
I just need to add a few tests (if a good samaritan wanted to contribute a few, I'd love that)

**Pull request closed without merging**